app/_gateway_entities/keyring.md

```diff
@@ -21,20 +21,103 @@ api_specs:
+When enabled, encryption and decryption of data are done on-the-fly by Kong immediately before writing to the database and after reading from the database. 
+Responses containing sensitive fields generated by the Admin API continue to show data as plain text. Kong runtime elements, such as plugins, can access sensitive fields transparently, without requiring additional configuration.
+
+## Encryption details
+
+### Encrypted fields
+
+The Keyring encrypts the following fields:
+
+- The `key` fields in a `certificate` object, which is the private key for a TLS certificate.
+- Certain configuration parameters in plugins and plugin-related authentication objects. These parameters have an `encrypted` label in the plugin's configuration reference.
+
+### Encryption and decryption
+
+{{site.base_gateway}} uses 256-bit AES encryption in GCM mode. Cryptographic nonce values for each encryption routine execution are derived from the kernel CSPRNG (`/dev/urandom`). The AES routines used by Kong are provided by the OpenSSL library bundled with the {{site.base_gateway}} package.
+
+## Key generation and lifecycle
+
+{{site.base_gateway}}’s keyring handling mechanisms allow for more than one key to be present on any given Kong node at a time. Each key may be used to read encrypted fields from the database, but only one key at any given time is used to write encrypted fields back to the data store. This process allows for a key rotation mechanism wherein new keyring material is introduced, and older keys may be present for a time to allow rotating previously-encrypted fields.
+
+Through the kernel CSPRNG, Kong derives keyring material generated through the `/keyring/generate` Admin API endpoint. Kong stores keyring material in a shared memory zone that all Kong worker processes access. To prevent key material from being written to disk as part of memory paging operations, we recommend that swap be disabled on systems running Kong.
+
+When operating in cluster mode, keyring material propagates automatically among all nodes in the Kong cluster. Because Kong nodes don't have a notion of direct peer-to-peer communication, the underlying data store serves as a communication channel to transmit messages. When a Kong node starts, it generates an ephemeral RSA key pair. The node’s public keys propagate to all other active nodes in the cluster. When an active node sees a message request for keyring material, it wraps the in-memory keyring material in the presented public key, and transmits the payload back over the central messaging channel provided by the underlying data store. This process allows each node in the cluster to broadcast keyring material to new nodes, without sending key material in plain text over the wire. This model requires that at least one node be running at all times within the cluster; a failure of all nodes requires manually re-importing the keyring to one node during an outage recovery.
```

Comment:
I'd like a diagram or a list that explicitly states what happens when keyring is enabled and data comes in that needs to be encrypted. It's difficult for me to process how keyring works with the long paragraphs of text.

What I'd like to see covered:
- Read/write encryption flow between the database
- how encrypted data is broadcasted to nodes
- How new/old keyring material is handled